### ISSUE TITLE
CLI is phenomenally slow to do anything at all

### ISSUE THREAD

**Issue by trejkaz** [crowdin/crowdin-cli-2#39]
The CLI was taking about 2 minutes to upload each individual file. We have about 300 files and 11 translations, which would take 300 × 11 × 2 = 6600 minutes, which is around 4.6 days.

So I thought I would run dtruss on the project to see how it's wasting all that time.

What it seems to be doing is calling stat64 on all the properties files over and over.

```
stat64("/Data/Projects/acme/trunk/product/investigator/src/resources/com/acme/investigator/workbench/items/nativeviewer/windows/vlc/Bundle_ar.properties\0", 0x700002445390, 0x84D00)		 = 0 0
stat64("/Data/Projects/acme/trunk/product/investigator/src/resources/com/acme/investigator/workbench/items/nativeviewer/windows/vlc/Bundle_da.properties\0", 0x700002445390, 0x84D00)		 = 0 0
stat64("/Data/Projects/acme/trunk/product/investigator/src/resources/com/acme/investigator/workbench/items/nativeviewer/windows/vlc/Bundle_de.properties\0", 0x700002445390, 0x84D00)		 = 0 0
stat64("/Data/Projects/acme/trunk/product/investigator/src/resources/com/acme/investigator/workbench/items/nativeviewer/windows/vlc/Bundle_es_419.properties\0", 0x700002445390, 0x84D00)		 = 0 0
stat64("/Data/Projects/acme/trunk/product/investigator/src/resources/com/acme/investigator/workbench/items/nativeviewer/windows/vlc/Bundle_fr_CA.properties\0", 0x700002445390, 0x84D00)		 = 0 0
stat64("/Data/Projects/acme/trunk/product/investigator/src/resources/com/acme/investigator/workbench/items/nativeviewer/windows/vlc/Bundle_iw.properties\0", 0x700002445390, 0x84D00)		 = 0 0
stat64("/Data/Projects/acme/trunk/product/investigator/src/resources/com/acme/investigator/workbench/items/nativeviewer/windows/vlc/Bundle_ja.properties\0", 0x700002445390, 0x84D00)		 = 0 0
stat64("/Data/Projects/acme/trunk/product/investigator/src/resources/com/acme/investigator/workbench/items/nativeviewer/windows/vlc/Bundle_ko.properties\0", 0x700002445390, 0x84D00)		 = 0 0
stat64("/Data/Projects/acme/trunk/product/investigator/src/resources/com/acme/investigator/workbench/items/nativeviewer/windows/vlc/Bundle_nl.properties\0", 0x700002445390, 0x84D00)		 = 0 0
stat64("/Data/Projects/acme/trunk/product/investigator/src/resources/com/acme/investigator/workbench/items/nativeviewer/windows/vlc/Bundle_pt_BR.properties\0", 0x700002445390, 0x84D00)		 = 0 0
stat64("/Data/Projects/acme/trunk/product/investigator/src/resources/com/acme/investigator/workbench/items/nativeviewer/windows/vlc/Bundle_zh_CN.properties\0", 0x700002445390, 0x84D00)		 = 0 0
stat64("/Data/Projects/acme/trunk/product/investigator/src/resources/com/acme/investigator/workbench/items/pdf/Bundle_ar.properties\0", 0x700002445390, 0x84D00)		 = 0 0
stat64("/Data/Projects/acme/trunk/product/investigator/src/resources/com/acme/investigator/workbench/items/pdf/Bundle_da.properties\0", 0x700002445390, 0x84D00)		 = 0 0
stat64("/Data/Projects/acme/trunk/product/investigator/src/resources/com/acme/investigator/workbench/items/pdf/Bundle_de.properties\0", 0x700002445390, 0x84D00)		 = 0 0
stat64("/Data/Projects/acme/trunk/product/investigator/src/resources/com/acme/investigator/workbench/items/pdf/Bundle_es_419.properties\0", 0x700002445390, 0x84D00)		 = 0 0
stat64("/Data/Projects/acme/trunk/product/investigator/src/resources/com/acme/investigator/workbench/items/pdf/Bundle_fr_CA.properties\0", 0x700002445390, 0x84D00)		 = 0 0
stat64("/Data/Projects/acme/trunk/product/investigator/src/resources/com/acme/investigator/workbench/items/pdf/Bundle_iw.properties\0", 0x700002445390, 0x84D00)		 = 0 0
stat64("/Data/Projects/acme/trunk/product/investigator/src/resources/com/acme/investigator/workbench/items/pdf/Bundle_ja.properties\0", 0x700002445390, 0x84D00)		 = 0 0
stat64("/Data/Projects/acme/trunk/product/investigator/src/resources/com/acme/investigator/workbench/items/pdf/Bundle_ko.properties\0", 0x700002445390, 0x84D00)		 = 0 0
stat64("/Data/Projects/acme/trunk/product/investigator/src/resources/com/acme/investigator/workbench/items/pdf/Bundle_nl.properties\0", 0x700002445390, 0x84D00)		 = 0 0
stat64("/Data/Projects/acme/trunk/product/investigator/src/resources/com/acme/investigator/workbench/items/pdf/Bundle_pt_BR.properties\0", 0x700002445390, 0x84D00)		 = 0 0
stat64("/Data/Projects/acme/trunk/product/investigator/src/resources/com/acme/investigator/workbench/items/pdf/Bundle_zh_CN.properties\0", 0x700002445390, 0x84D00)		 = 0 0
stat64("/Data/Projects/acme/trunk/product/investigator/src/resources/com/acme/investigator/workbench/items/printpreview/Bundle_ar.properties\0", 0x700002445390, 0x84D00)		 = 0 0
```

I tried grepping one particular file to figure out how often it is doing this, and it prints about one line per second. So every second, it's stating every file in the directory hierarchy.

I haven't read the code yet, so I can't figure out why it would have to scan the whole directory so many times before uploading just one file every couple of minutes.

**Comment by trejkaz** [comment]
JVisualVM shows isFile() being called over and over and over being the main time sink.

CSV inline since GitHub won't let me attach it.

{noformat}
"Call Tree - Method","Total Time [%]","Total Time","Total Time (CPU)","Samples",
" All threads","0.0","0","0","0"
"  main","100.0","59598941","59598941","1"
"   com.crowdin.cli.Cli.main()","100.0","59598941","59598941","1"
"    com.crowdin.cli.commands.Commands.run()","100.0","59598941","59598941","1"
"     com.crowdin.cli.commands.Commands.uploadTranslation()","100.0","59598941","59598941","1"
"      com.crowdin.cli.utils.CommandUtils.doLanguagesMapping()","100.0","59598941","59598941","1"
"       com.crowdin.cli.utils.CommandUtils.getSourcesWithoutIgnores()","100.0","59598941","59598941","1"
"        java.io.File.isFile()","86.92","51803397","51803397","3"
"        com.crowdin.cli.properties.helper.FileHelper.getIgnoreFiles()","10.57349","6301688","6301688","2"
"         com.crowdin.cli.properties.helper.FileHelper.filterFiles()","10.57349","6301688","6301688","20"
"          com.crowdin.cli.properties.helper.FileHelper.getlistDirectory()","6.4931407","3869843","3869843","10"
"           com.crowdin.cli.properties.helper.FileHelper.getlistDirectory()","6.4931407","3869843","3869843","10"
"            com.crowdin.cli.properties.helper.FileHelper.getlistDirectory()","6.4931407","3869843","3869843","10"
"             com.crowdin.cli.properties.helper.FileHelper.getlistDirectory()","6.4931407","3869843","3869843","10"
"              com.crowdin.cli.properties.helper.FileHelper.getlistDirectory()","6.04225","3601117","3601117","10"
"               com.crowdin.cli.properties.helper.FileHelper.getlistDirectory()","5.5404305","3302038","3302038","12"
"                com.crowdin.cli.properties.helper.FileHelper.getlistDirectory()","5.0332437","2999760","2999760","11"
"                 com.crowdin.cli.properties.helper.FileHelper.getlistDirectory()","4.531933","2700984","2700984","12"
"                  com.crowdin.cli.properties.helper.FileHelper.getlistDirectory()","3.69521","2202306","2202306","11"
"                   com.crowdin.cli.properties.helper.FileHelper.getlistDirectory()","2.5249057","1504817","1504817","10"
"                    java.io.File.listFiles()","1.1806301","703643","703643","5"
"                    com.crowdin.cli.properties.helper.FileHelper.getlistDirectory()","1.0032024","597898","597898","5"
"                     com.crowdin.cli.properties.helper.FileHelper.getlistDirectory()","0.49578398","295482","295482","3"
"                      com.crowdin.cli.properties.helper.FileHelper.getlistDirectory()","0.33480963","199543","199543","2"
"                       com.crowdin.cli.properties.helper.FileHelper.getlistDirectory()","0.33480963","199543","199543","2"
"                        java.io.File.listFiles()","0.16766238","99925","99925","1"
"                        com.crowdin.cli.properties.helper.FileHelper.getlistDirectory()","0.16714558","99617","99617","1"
"                         java.io.File.isDirectory()","0.16714558","99617","99617","1"
"                         Self time","0.0","0","0","1"
"                        Self time","0.0","0","0","2"
"                       Self time","0.0","0","0","2"
"                      java.io.File.isDirectory()","0.16097434","95939","95939","1"
"                      Self time","0.0","0","0","3"
"                     java.io.File.listFiles()","0.33803117","201463","201463","2"
"                     java.io.File.isDirectory()","0.16938557","100952","100952","1"
"                     Self time","0.0","0","0","5"
"                    java.io.File.isDirectory()","0.3410715","203275","203275","2"
"                    Self time","0.0","0","0","10"
"                   java.io.File.isDirectory()","0.8308134","495156","495156","5"
"                   java.io.File.listFiles()","0.33948925","202332","202332","2"
"                   Self time","0.0","0","0","11"
"                  java.io.File.isDirectory()","0.6707418","399755","399755","4"
"                  java.io.File.listFiles()","0.16597778","98921","98921","1"
"                  Self time","0.0","0","0","12"
"                 java.io.File.isDirectory()","0.33343378","198723","198723","2"
"                 java.io.File.listFiles()","0.16787714","100053","100053","1"
"                 Self time","0.0","0","0","11"
"                java.io.File.listFiles()","0.33930635","202223","202223","2"
"                java.io.File.isDirectory()","0.16787882","100054","100054","1"
"                Self time","0.0","0","0","12"
"               java.io.File.isDirectory()","0.33695564","200822","200822","2"
"               java.io.File.listFiles()","0.16486199","98256","98256","1"
"               Self time","0.0","0","0","10"
"              java.io.File.listFiles()","0.45089057","268726","268726","3"
"              Self time","0.0","0","0","10"
"             Self time","0.0","0","0","10"
"            Self time","0.0","0","0","10"
"           Self time","0.0","0","0","10"
"          java.io.File.listFiles()","3.915927","2333851","2333851","10"
"          java.io.File.isDirectory()","0.16441903","97992","97992","1"
"          Self time","0.0","0","0","20"
"         Self time","0.0","0","0","2"
"        com.crowdin.cli.properties.helper.FileHelper.getFileSource()","2.5065143","1493856","1493856","2"
"         com.crowdin.cli.properties.helper.FileHelper.filterFiles()","2.5065143","1493856","1493856","6"
"          com.crowdin.cli.properties.helper.FileHelper.getlistDirectory()","1.6790953","1000723","1000723","4"
"           com.crowdin.cli.properties.helper.FileHelper.getlistDirectory()","1.6790953","1000723","1000723","4"
"            com.crowdin.cli.properties.helper.FileHelper.getlistDirectory()","1.6790953","1000723","1000723","4"
"             com.crowdin.cli.properties.helper.FileHelper.getlistDirectory()","1.6790953","1000723","1000723","4"
"              com.crowdin.cli.properties.helper.FileHelper.getlistDirectory()","1.5115789","900885","900885","3"
"               com.crowdin.cli.properties.helper.FileHelper.getlistDirectory()","1.3474048","803039","803039","2"
"                com.crowdin.cli.properties.helper.FileHelper.getlistDirectory()","1.0069709","600144","600144","2"
"                 com.crowdin.cli.properties.helper.FileHelper.getlistDirectory()","0.8318789","495791","495791","3"
"                  com.crowdin.cli.properties.helper.FileHelper.getlistDirectory()","0.5037321","300219","300219","3"
"                   com.crowdin.cli.properties.helper.FileHelper.getlistDirectory()","0.17204668","102538","102538","1"
"                    java.io.File.listFiles()","0.17204668","102538","102538","1"
"                    Self time","0.0","0","0","1"
"                   java.io.File.isDirectory()","0.16799627","100124","100124","1"
"                   java.io.File.listFiles()","0.16368747","97556","97556","1"
"                   Self time","0.0","0","0","3"
"                  java.io.File.isDirectory()","0.3281451","195571","195571","2"
"                  Self time","0.0","0","0","3"
"                 java.io.File.listFiles()","0.17509204","104353","104353","1"
"                 Self time","0.0","0","0","2"
"                java.io.File.listFiles()","0.3404339","202895","202895","2"
"                Self time","0.0","0","0","2"
"               java.io.File.isDirectory()","0.16417238","97845","97845","1"
"               Self time","0.0","0","0","3"
"              java.io.File.listFiles()","0.16751473","99837","99837","1"
"              Self time","0.0","0","0","4"
"             Self time","0.0","0","0","4"
"            Self time","0.0","0","0","4"
"           Self time","0.0","0","0","4"
"          java.io.File.listFiles()","0.664146","395824","395824","2"
"          java.io.File.isDirectory()","0.16327135","97308","97308","1"
"          Self time","0.0","0","0","6"
"         Self time","0.0","0","0","2"
"        Self time","0.0","0","0","1"
"       Self time","0.0","0","0","1"
"      Self time","0.0","0","0","1"
"     Self time","0.0","0","0","1"
"    Self time","0.0","0","0","1"
{noformat}

**Comment by trejkaz** [comment]
What I think the problem is from some browsing of the code:

* In `CommandUtils#getTranslations`, you're supposed to be getting the translations for just one file, but you're actually calling `getSourcesWithoutIgnores`, which iterates every single file.
* `getTranslations` is then called once per source file.
* `getTranslations` is operating over the result of `getSourcesWithoutIgnores`, so it is even looking up translations for ignored files, which wastes even more time.
* `isFile()` is being called a lot for every file when you already know it exists because some other code returned it by listing the directory.
* `isFile()` is even being called for ignored files.

Additionally, computing the actual ignore list in `getSourcesWithoutIgnores` is very slow because it is iterating two lists:

https://github.com/crowdin/crowdin-cli-2/blob/master/src/main/java/com/crowdin/cli/utils/CommandUtils.java#L125

Set subtraction or even set `contains()` calls would be much faster.

In short, I had cloned the project to see whether I could fix it, but there are simply too many problems to fix and I feel that I would end up with a completely different program if I tried. So I think initially I will have to write off the CLI as not a workable solution for the time being.

**Comment by ihor-popyk** [comment]
How do you specify the support team that source files uploaded ~ 7 minutes, translations about the same. For a number of files consider this time is acceptable.
About isFile() checks - agree. Already made some changes (#44) and the next version will be faster.
Thank you very much for the work.

**Comment by trejkaz** [comment]
I updated to version 2.0.16 and it's still extremely slow to do `upload translations`. I have waited an hour so far, and it hasn't printed a single line yet.

**Comment by trejkaz** [comment]
Initial investigation shows that files I have asked the tool to exclude are still being read.

Initial investigation shows that files I have asked the tool to exclude are still being read.

```
stat64("/Data/Projects/acme/trunk/product/release/macosx/desktop/build/Acme/Applications/Keynote.app/Contents/Resources/Frames/Playtime Waves.sfrframe/Contents/Resources/Playtime_waves_01.png\0", 0x70000F30CAA0, 0x1000)		 = 0 0
stat64("/Data/Projects/acme/trunk/product/release/macosx/desktop/build/Acme/Applications/Keynote.app/Contents/Resources/Frames/Playtime Waves.sfrframe/Contents/Resources/Playtime_waves_02.png\0", 0x70000F30CAA0, 0x1000)		 = 0 0
stat64("/Data/Projects/acme/trunk/product/release/macosx/desktop/build/Acme/Applications/Keynote.app/Contents/Resources/Frames/Playtime Waves.sfrframe/Contents/Resources/Playtime_waves_03.png\0", 0x70000F30CAA0, 0x1000)		 = 0 0
stat64("/Data/Projects/acme/trunk/product/release/macosx/desktop/build/Acme/Applications/Keynote.app/Contents/Resources/Frames/Playtime Waves.sfrframe/Contents/Resources/Playtime_waves_04.png\0", 0x70000F30CAA0, 0x1000)		 = 0 0
stat64("/Data/Projects/acme/trunk/product/release/macosx/desktop/build/Acme/Applications/Keynote.app/Contents/Resources/Frames/Playtime Waves.sfrframe/Contents/Resources/Playtime_waves_06.png\0", 0x70000F30CAA0, 0x1000)		 = 0 0
stat64("/Data/Projects/acme/trunk/product/release/macosx/desktop/build/Acme/Applications/Keynote.app/Contents/Resources/Frames/Playtime Waves.sfrframe/Contents/Resources/Playtime_waves_07.png\0", 0x70000F30CAA0, 0x1000)		 = 0 0
stat64("/Data/Projects/acme/trunk/product/release/macosx/desktop/build/Acme/Applications/Keynote.app/Contents/Resources/Frames/Playtime Waves.sfrframe/Contents/Resources/Playtime_waves_08.png\0", 0x70000F30CAA0, 0x1000)		 = 0 0
stat64("/Data/Projects/acme/trunk/product/release/macosx/desktop/build/Acme/Applications/Keynote.app/Contents/Resources/Frames/Playtime Waves.sfrframe/Contents/Resources/Playtime_waves_09.png\0", 0x70000F30CAA0, 0x1000)		 = 0 0
stat64("/Data/Projects/acme/trunk/product/release/macosx/desktop/build/Acme/Applications/Keynote.app/Contents/Resources/Frames/Playtime Waves.sfrframe/Contents/Resources/wellImage.png\0", 0x70000F30CAA0, 0x1000)		 = 0 0
stat64("/Data/Projects/acme/trunk/product/release/macosx/desktop/build/Acme/Applications/Keynote.app/Contents/Resources/Frames/Playtime Waves.sfrframe/Contents/Resources/wellImage@2x.png\0", 0x70000F30CAA0, 0x1000)		 = 0 0
```

So there's two issues going on here:

1) The "ignore" setting is still not working - if it were, we wouldn't be seeing files with /build/ in them being accessed.

2) The "Applications" symlink to "/Applications" is being followed, and then all of my installed applications are being searched for resource bundles.

**Comment by trejkaz** [comment]
Initial look at the path matching code - for some reason, this is being implemented from scratch, whereas Java already has `FileSystems.getDefault().getPathMatcher(String)` which seems to implement all the documented syntax, plus more.

I took a shot at converting to use this `PathMatcher`, but there are problems:

* There are no tests, so I have no way to verify whether I'm breaking anything.
* There are things being done in the code which are not part of the documented syntax. e.g.:
    * Regex support
    * Path elements ending with a colon are treated differently and there is no comment explaining why

If you do have a way to test whether it's doing the right thing, I would suggest trying to slot in `FileSystems.getDefault().getPathMatcher("glob:" + pattern)` and see whether it supports all the supported syntax correctly.

For the symlink side of things... using the newer `Path` API instead of the legacy `File` API might be the way forward for that too, because it usually has parameters to tell it not to follow links.
